io/src/csv/read_cells.cpp

```diff
@@ -96,9 +97,26 @@ void read_cells(
     std::vector<std::pair<csv::cell, unsigned int>> allCells;
     allCells.reserve(50000);
 
+    // Keep track of duplicate cells
+    using cell_key_t =
+        std::tuple<decltype(std::declval<csv::cell>().geometry_id),
+                   decltype(std::declval<csv::cell>().channel0),
+                   decltype(std::declval<csv::cell>().channel1)>;
+
+    std::set<cell_key_t> seen;
+    std::size_t dedup = 0;
+
     // Read all cells from input file.
     csv::cell iocell;
     while (reader.read(iocell)) {
+        cell_key_t key = {iocell.geometry_id, iocell.channel0, iocell.channel1};
+
+        if (seen.find(key) != seen.end()) {
+            dedup++;
```

Comment:
Let's add a check here that their activation values would be ~the same. If they are not, we're dealing with an actual error, and the code should throw an exception about it. But for "real duplicates" this can indeed be a good idea to suppress them with a warning.

Reply:
So what I suspect is happening here is that we simulate particles, they cross some surface around the same spot, and a cell is recorded for each particle, possibly in the same pixel. It makes sense, then, that the activation values could be quite different. So I don't think we'll find my cases where the activation value is the same.

I'm not sure how this kind of detector response simulation should behave, e.g. do two particles depositing energy in a pixel simply add up? Or is there some more complex way to combine them?

@beomki-yeo you designed this test, perhaps you can weigh in?

Reply:
The test seems to be designed specifically to test cases in which two particles activate the same cell, so I think simply deduplicating the input file will defeat the purpose of the test. :frowning:

Reply:
But then the test is bad... :thinking: The real hardware will never report the same pixel multiple times, unless there's a bug somewhere in the data readout chain. So the simulation shouldn't behave differently either.

Yes, the deposited energies generally simply add up. Naively, you would first collect all energy deposits into some in-memory container, and then go through the pixels one by one, adding up all deposits that happened in the geometric volume of that pixel. With a simple addition. (Energy can be as simple as that. :wink:)

It's not really a debate whether duplicate cells in the CSV files are bad or not. :thinking: We can certainly discuss about the ordering of the cells (both in the files and in memory). But duplicates are definitely a no-no.

Reply:
For energy deposit for multiple particles on the same cell we might need an expert.  But we can just  add up as you suggested for the moment.

> I think simply deduplicating the input file will defeat the purpose of the test

Right. I think we need to leave `event_map` (truth) as it is and just shrinking the cell readout data. This should be possible I guess?

Reply:
> We can certainly discuss about the ordering of the cells (both in the files and in memory). But duplicates are definitely a no-no

I think we need to consider splitting cell data and particle-to-cell data (Like there is a dedicated map for measurement-to-hit data)

Reply:
:confused: In the end, once every one of our files is set up correctly, we should have:
  - particles pointing at a list of cells that they "produced";
  - cells pointing at a list of particles that contributed to them.

A cell is still just a single thing, I don't see that changing. Our analysis model has to be able to deal with multiple particles contributing to the same cell. I can't imagine that the analysis done in Athena would happen in any other way. :thinking: